src/ext/php7/engine_hooks.c

```diff
@@ -194,7 +269,87 @@ static bool _dd_should_trace_call(zend_execute_data *call, zend_function *fbc, d
+static void dd_copy_prehook_args(zval *args, zend_execute_data *call) {
+    uint32_t i, first_extra_arg, extra_arg_count;
+    zval *p, *q;
+    uint32_t arg_count = ZEND_CALL_NUM_ARGS(call);
+
+    // @see https://github.com/php/php-src/blob/PHP-7.0/Zend/zend_builtin_functions.c#L506-L562
+    array_init_size(args, arg_count);
+    if (arg_count && call->func) {
+        first_extra_arg = call->func->op_array.num_args;
+        bool has_extra_args = arg_count > first_extra_arg;
+
+        zend_hash_real_init(Z_ARRVAL_P(args), 1);
+        ZEND_HASH_FILL_PACKED(Z_ARRVAL_P(args)) {
+            i = 0;
+            p = ZEND_CALL_ARG(call, 1);
+            if (has_extra_args) {
+                i = arg_count - first_extra_arg;
+            }
+            while (i < arg_count) {
+                q = p;
+                DD_TRACE_COPY_NULLABLE_ARG(q);
+                p++;
+                i++;
+            }
+            /* If we are copying arguments before i_init_func_execute_data() has run, the extra agruments
+               have not yet been moved to a separate array. */
+            if (has_extra_args) {
+                p = ZEND_CALL_VAR_NUM(call, first_extra_arg);
+                extra_arg_count = arg_count - first_extra_arg;
+                while (extra_arg_count--) {
+                    q = p;
+                    DD_TRACE_COPY_NULLABLE_ARG(q);
+                    p++;
+                }
+            }
+        }
+        ZEND_HASH_FILL_END();
+        Z_ARRVAL_P(args)->nNumOfElements = arg_count;
+    }
+}
+
+static void dd_copy_posthook_args(zval *args, zend_execute_data *call) {
+    uint32_t i, first_extra_arg;
+    zval *p, *q;
+    uint32_t arg_count = ZEND_CALL_NUM_ARGS(call);
+
+    // @see https://github.com/php/php-src/blob/PHP-7.0/Zend/zend_builtin_functions.c#L506-L562
+    array_init_size(args, arg_count);
+    if (arg_count && call->func) {
+        first_extra_arg = call->func->op_array.num_args;
+        bool has_extra_args = arg_count > first_extra_arg;
+
+        zend_hash_real_init(Z_ARRVAL_P(args), 1);
+        ZEND_HASH_FILL_PACKED(Z_ARRVAL_P(args)) {
+            i = 0;
+            p = ZEND_CALL_ARG(call, 1);
+            if (has_extra_args) {
+                while (i < first_extra_arg) {
+                    q = p;
+                    DD_TRACE_COPY_NULLABLE_ARG(q);
+                    p++;
+                    i++;
+                }
+                if (call->func->type != ZEND_INTERNAL_FUNCTION) {
+                    p = ZEND_CALL_VAR_NUM(call, call->func->op_array.last_var + call->func->op_array.T);
+                }
+            }
+            while (i < arg_count) {
+                q = p;
+                DD_TRACE_COPY_NULLABLE_ARG(q);
+                p++;
+                i++;
+            }
+        }
+        ZEND_HASH_FILL_END();
+        Z_ARRVAL_P(args)->nNumOfElements = arg_count;
+    }
+}
+
+#if 0
```

Comment:
Not a blocker for merge, but is there a reason to keep the old copy args around?